presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -183,6 +184,155 @@ else if (rescaleFactor > 0) {
         }
     }
 
+    public static Slice add(Slice left, Slice right)
+    {
+        Slice result = unscaledDecimal();
+
+        if (!(isNegative(left) ^ isNegative(right))) {
+            // either both negative or both positive
+            addUnsigned(left, right, result, isNegative(left));
+        }
+        else {
+            int compare = compareUnsigned(left, right);
+            if (compare == 0) {
+                return result;
+            }
+            if (isNegative(left)) {
```

Comment:
I think you can extract isNegative as separate boolean and put it to `subtractUnsigned` with possible negation (depends if `compare > 0`). Then you avoid this branch